File: packages/js-sdk/src/modules/partner.ts
/**
 * TurboPartner Module - Partner portal operations
 */

import { HttpClient, PartnerClientConfig } from '../http';
import {
  CreateOrganizationRequest,
  ListOrganizationsRequest,
  UpdateOrganizationRequest,
  UpdateEntitlementsRequest,
  AddOrgUserRequest,
  ListOrgUsersRequest,
  UpdateOrgUserRequest,
  CreateOrgApiKeyRequest,
  ListOrgApiKeysRequest,
  UpdateOrgApiKeyRequest,
  CreatePartnerApiKeyRequest,
  ListPartnerApiKeysRequest,
  UpdatePartnerApiKeyRequest,
  AddPartnerUserRequest,
  ListPartnerUsersRequest,
  UpdatePartnerUserRequest,
  ListAuditLogsRequest,
  SuccessResponse,
  OrganizationResponse,
  OrganizationListResponse,
  OrganizationDetailResponse,
  EntitlementsResponse,
  OrgUserResponse,
  OrgUserListResponse,
  OrgApiKeyResponse,
  OrgApiKeyListResponse,
  PartnerApiKeyResponse,
  PartnerApiKeyListResponse,
  PartnerUserResponse,
  PartnerUserListResponse,
  AuditLogListResponse,
} from '../types/partner';

/**
 * Convert a request object to query parameters, filtering out undefined/null values.
 * Booleans are serialized as 'true'/'false' strings.
 */
function toQueryParams(request?: Record<string, any>): Record<string, string> | undefined {
  if (!request) return undefined;

  const params: Record<string, string> = {};
  for (const [key, value] of Object.entries(request)) {
    if (value !== undefined && value !== null) {
      if (typeof value === 'boolean') {
        params[key] = value ? 'true' : 'false';
      } else {
        params[key] = String(value);
      }
    }
  }
  return Object.keys(params).length > 0 ? params : undefined;
}

export class TurboPartner {
  private static client: HttpClient;
  private static partnerId: string;

  /**
   * Configure the TurboPartner module with partner API credentials
   *
   * @param config - Partner configuration object
   * @param config.partnerApiKey - Partner API key (must start with TDXP-)
   * @param config.partnerId - Partner ID (UUID format)
   * @param config.baseUrl - API base URL (optional, defaults to https://api.turbodocx.com)
   *
   * @example
   * ```typescript
   * TurboPartner.configure({
   *   partnerApiKey: process.env.TURBODOCX_PARTNER_API_KEY,
   *   partnerId: process.env.TURBODOCX_PARTNER_ID,
   * });
   * ```
   */
  static configure(config: PartnerClientConfig): void {
    this.client = new HttpClient({
      apiKey: config.partnerApiKey,
      baseUrl: config.baseUrl,
      skipSenderValidation: true,
    });
    this.partnerId = config.partnerId;
  }

  /**
   * Get the HTTP client instance, initializing if necessary
   */
  private static getClient(): HttpClient {
    if (!this.client) {
      const partnerApiKey = process.env.TURBODOCX_PARTNER_API_KEY;
      const partnerId = process.env.TURBODOCX_PARTNER_ID;
      if (!partnerApiKey || !partnerId) {
        throw new Error('TurboPartner must be configured before use. Call TurboPartner.configure() or set TURBODOCX_PARTNER_API_KEY and TURBODOCX_PARTNER_ID environment variables.');
      }
      this.configure({ partnerApiKey, partnerId });
    }
    return this.client;
  }

  /**
   * Get the partner ID
   */
  private static getPartnerId(): string {
    if (!this.partnerId) {
      // Trigger auto-initialization
      this.getClient();
    }
    return this.partnerId;
  }

  // ============================================
  // ORGANIZATION MANAGEMENT
  // ============================================

  /**
   * Create a new organization
   *
   * @param request - Organization creation request
   * @returns Created organization details
   *
   * @example
   * ```typescript
   * const org = await TurboPartner.createOrganization({
   *   name: 'Acme Corp',
   *   metadata: { industry: 'Technology' },
   * });
   * console.log(org.data.id);
   * ```
   */
  static async createOrganization(request: CreateOrganizationRequest): Promise<OrganizationResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.post<OrganizationResponse>(`/partner/${partnerId}/organization`, request);
  }

  /**
   * List all organizations
   *
   * @param request - Optional pagination and search parameters
   * @returns Paginated list of organizations
   *
   * @example
   * ```typescript
   * const orgs = await TurboPartner.listOrganizations({ limit: 10, search: 'acme' });
   * console.log(orgs.data.results);
   * ```
   */
  static async listOrganizations(request?: ListOrganizationsRequest): Promise<OrganizationListResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.get<OrganizationListResponse>(`/partner/${partnerId}/organizations`, toQueryParams(request));
  }

  /**
   * Get detailed information about an organization
   *
   * @param organizationId - Organization ID
   * @returns Organization details including features and tracking
   *
   * @example
   * ```typescript
   * const details = await TurboPartner.getOrganizationDetails('org-uuid');
   * console.log(details.data.features);
   * ```
   */
  static async getOrganizationDetails(organizationId: string): Promise<OrganizationDetailResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.get<OrganizationDetailResponse>(`/partner/${partnerId}/organizations/${organizationId}`);
  }

  /**
   * Update organization information
   *
   * @param organizationId - Organization ID
   * @param request - Update request with new name
   * @returns Updated organization details
   *
   * @example
   * ```typescript
   * const org = await TurboPartner.updateOrganizationInfo('org-uuid', { name: 'New Name' });
   * ```
   */
  static async updateOrganizationInfo(organizationId: string, request: UpdateOrganizationRequest): Promise<OrganizationResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.patch<OrganizationResponse>(`/partner/${partnerId}/organizations/${organizationId}`, request);
  }

  /**
   * Delete an organization
   *
   * @param organizationId - Organization ID
   * @returns Success response
   *
   * @example
   * ```typescript
   * await TurboPartner.deleteOrganization('org-uuid');
   * ```
   */
  static async deleteOrganization(organizationId: string): Promise<SuccessResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.delete<SuccessResponse>(`/partner/${partnerId}/organizations/${organizationId}`);
  }

  /**
   * Update organization entitlements (features and tracking)
   *
   * @param organizationId - Organization ID
   * @param request - Entitlements update request
   * @returns Updated entitlements
   *
   * @example
   * ```typescript
   * const result = await TurboPartner.updateOrganizationEntitlements('org-uuid', {
   *   features: { maxUsers: 50, hasTDAI: true },
   * });
   * ```
   */
  static async updateOrganizationEntitlements(organizationId: string, request: UpdateEntitlementsRequest): Promise<EntitlementsResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.patch<EntitlementsResponse>(`/partner/${partnerId}/organizations/${organizationId}/entitlements`, request);
  }

  // ============================================
  // ORGANIZATION USER MANAGEMENT
  // ============================================

  /**
   * List users in an organization
   *
   * @param organizationId - Organization ID
   * @param request - Optional pagination and search parameters
   * @returns Paginated list of organization users
   *
   * @example
   * ```typescript
   * const users = await TurboPartner.listOrganizationUsers('org-uuid', { limit: 25 });
   * ```
   */
  static async listOrganizationUsers(organizationId: string, request?: ListOrgUsersRequest): Promise<OrgUserListResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.get<OrgUserListResponse>(`/partner/${partnerId}/organizations/${organizationId}/users`, toQueryParams(request));
  }

  /**
   * Add a user to an organization
   *
   * @param organizationId - Organization ID
   * @param request - User details with email and role
   * @returns Created user details
   *
   * @example
   * ```typescript
   * const user = await TurboPartner.addUserToOrganization('org-uuid', {
   *   email: 'user@example.com',
   *   role: 'contributor',
   * });
   * ```
   */
  static async addUserToOrganization(organizationId: string, request: AddOrgUserRequest): Promise<OrgUserResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.post<OrgUserResponse>(`/partner/${partnerId}/organizations/${organizationId}/users`, request);
  }

  /**
   * Update a user's role in an organization
   *
   * @param organizationId - Organization ID
   * @param userId - User ID
   * @param request - Updated role
   * @returns Updated user details
   *
   * @example
   * ```typescript
   * const user = await TurboPartner.updateOrganizationUserRole('org-uuid', 'user-uuid', {
   *   role: 'admin',
   * });
   * ```
   */
  static async updateOrganizationUserRole(organizationId: string, userId: string, request: UpdateOrgUserRequest): Promise<OrgUserResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.patch<OrgUserResponse>(`/partner/${partnerId}/organizations/${organizationId}/users/${userId}`, request);
  }

  /**
   * Remove a user from an organization
   *
   * @param organizationId - Organization ID
   * @param userId - User ID
   * @returns Success response
   *
   * @example
   * ```typescript
   * await TurboPartner.removeUserFromOrganization('org-uuid', 'user-uuid');
   * ```
   */
  static async removeUserFromOrganization(organizationId: string, userId: string): Promise<SuccessResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.delete<SuccessResponse>(`/partner/${partnerId}/organizations/${organizationId}/users/${userId}`);
  }

  /**
   * Resend an organization invitation email to a user
   *
   * @param organizationId - Organization ID
   * @param userId - User ID
   * @returns Success response
   *
   * @example
   * ```typescript
   * await TurboPartner.resendOrganizationInvitationToUser('org-uuid', 'user-uuid');
   * ```
   */
  static async resendOrganizationInvitationToUser(organizationId: string, userId: string): Promise<SuccessResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.post<SuccessResponse>(`/partner/${partnerId}/organizations/${organizationId}/users/${userId}/resend-invitation`);
  }

  // ============================================
  // ORGANIZATION API KEY MANAGEMENT
  // ============================================

  /**
   * List API keys for an organization
   *
   * @param organizationId - Organization ID
   * @param request - Optional pagination and search parameters
   * @returns Paginated list of API keys
   *
   * @example
   * ```typescript
   * const keys = await TurboPartner.listOrganizationApiKeys('org-uuid');
   * ```
   */
  static async listOrganizationApiKeys(organizationId: string, request?: ListOrgApiKeysRequest): Promise<OrgApiKeyListResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.get<OrgApiKeyListResponse>(`/partner/${partnerId}/organizations/${organizationId}/apikeys`, toQueryParams(request));
  }

  /**
   * Create an API key for an organization
   *
   * @param organizationId - Organization ID
   * @param request - API key creation request with name and role
   * @returns Created API key (includes the full key value only on creation)
   *
   * @example
   * ```typescript
   * const key = await TurboPartner.createOrganizationApiKey('org-uuid', {
   *   name: 'Production Key',
   *   role: 'admin',
   * });
   * console.log(key.data.key); // Full key only shown on creation
   * ```
   */
  static async createOrganizationApiKey(organizationId: string, request: CreateOrgApiKeyRequest): Promise<OrgApiKeyResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.post<OrgApiKeyResponse>(`/partner/${partnerId}/organizations/${organizationId}/apikeys`, request);
  }

  /**
   * Update an organization API key
   *
   * @param organizationId - Organization ID
   * @param apiKeyId - API key ID
   * @param request - Updated key properties
   * @returns Updated API key details
   *
   * @example
   * ```typescript
   * const key = await TurboPartner.updateOrganizationApiKey('org-uuid', 'key-uuid', {
   *   name: 'Updated Key Name',
   * });
   * ```
   */
  static async updateOrganizationApiKey(organizationId: string, apiKeyId: string, request: UpdateOrgApiKeyRequest): Promise<OrgApiKeyResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.patch<OrgApiKeyResponse>(`/partner/${partnerId}/organizations/${organizationId}/apikeys/${apiKeyId}`, request);
  }

  /**
   * Revoke an organization API key
   *
   * @param organizationId - Organization ID
   * @param apiKeyId - API key ID
   * @returns Success response
   *
   * @example
   * ```typescript
   * await TurboPartner.revokeOrganizationApiKey('org-uuid', 'key-uuid');
   * ```
   */
  static async revokeOrganizationApiKey(organizationId: string, apiKeyId: string): Promise<SuccessResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.delete<SuccessResponse>(`/partner/${partnerId}/organizations/${organizationId}/apikeys/${apiKeyId}`);
  }

  // ============================================
  // PARTNER API KEY MANAGEMENT
  // ============================================

  /**
   * List partner API keys
   *
   * @param request - Optional pagination and search parameters
   * @returns Paginated list of partner API keys
   *
   * @example
   * ```typescript
   * const keys = await TurboPartner.listPartnerApiKeys({ limit: 10 });
   * ```
   */
  static async listPartnerApiKeys(request?: ListPartnerApiKeysRequest): Promise<PartnerApiKeyListResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.get<PartnerApiKeyListResponse>(`/partner/${partnerId}/api-keys`, toQueryParams(request));
  }

  /**
   * Create a new partner API key
   *
   * @param request - API key creation request with name and scopes
   * @returns Created API key (includes the full key value only on creation)
   *
   * @example
   * ```typescript
   * const key = await TurboPartner.createPartnerApiKey({
   *   name: 'CI/CD Key',
   *   scopes: ['org:create', 'org:read'],
   *   description: 'Key for automated deployments',
   * });
   * console.log(key.data.key); // Full key only shown on creation
   * ```
   */
  static async createPartnerApiKey(request: CreatePartnerApiKeyRequest): Promise<PartnerApiKeyResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.post<PartnerApiKeyResponse>(`/partner/${partnerId}/api-keys`, request);
  }

  /**
   * Update a partner API key
   *
   * @param keyId - API key ID
   * @param request - Updated key properties
   * @returns Updated API key details
   *
   * @example
   * ```typescript
   * const key = await TurboPartner.updatePartnerApiKey('key-uuid', {
   *   name: 'Updated Name',
   *   scopes: ['org:create', 'org:read', 'org:update'],
   * });
   * ```
   */
  static async updatePartnerApiKey(keyId: string, request: UpdatePartnerApiKeyRequest): Promise<PartnerApiKeyResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.patch<PartnerApiKeyResponse>(`/partner/${partnerId}/api-keys/${keyId}`, request);
  }

  /**
   * Revoke a partner API key
   *
   * @param keyId - API key ID
   * @returns Success response
   *
   * @example
   * ```typescript
   * await TurboPartner.revokePartnerApiKey('key-uuid');
   * ```
   */
  static async revokePartnerApiKey(keyId: string): Promise<SuccessResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.delete<SuccessResponse>(`/partner/${partnerId}/api-keys/${keyId}`);
  }

  // ============================================
  // PARTNER USER MANAGEMENT
  // ============================================

  /**
   * List partner portal users
   *
   * @param request - Optional pagination and search parameters
   * @returns Paginated list of partner users
   *
   * @example
   * ```typescript
   * const users = await TurboPartner.listPartnerPortalUsers({ limit: 25 });
   * ```
   */
  static async listPartnerPortalUsers(request?: ListPartnerUsersRequest): Promise<PartnerUserListResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.get<PartnerUserListResponse>(`/partner/${partnerId}/users`, toQueryParams(request));
  }

  /**
   * Add a user to the partner portal
   *
   * @param request - User details with email, role, and permissions
   * @returns Created user details
   *
   * @example
   * ```typescript
   * const user = await TurboPartner.addUserToPartnerPortal({
   *   email: 'admin@partner.com',
   *   role: 'admin',
   *   permissions: {
   *     canManageOrgs: true,
   *     canManageOrgUsers: true,
   *     canManagePartnerUsers: false,
   *     canManageOrgAPIKeys: true,
   *     canManagePartnerAPIKeys: false,
   *     canUpdateEntitlements: true,
   *     canViewAuditLogs: true,
   *   },
   * });
   * ```
   */
  static async addUserToPartnerPortal(request: AddPartnerUserRequest): Promise<PartnerUserResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.post<PartnerUserResponse>(`/partner/${partnerId}/users`, request);
  }

  /**
   * Update a partner user's role and permissions
   *
   * @param userId - User ID
   * @param request - Updated role and/or permissions
   * @returns Updated user details
   *
   * @example
   * ```typescript
   * const user = await TurboPartner.updatePartnerUserPermissions('user-uuid', {
   *   role: 'member',
   *   permissions: { canManageOrgs: true },
   * });
   * ```
   */
  static async updatePartnerUserPermissions(userId: string, request: UpdatePartnerUserRequest): Promise<PartnerUserResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.patch<PartnerUserResponse>(`/partner/${partnerId}/users/${userId}`, request);
  }

  /**
   * Remove a user from the partner portal
   *
   * @param userId - User ID
   * @returns Success response
   *
   * @example
   * ```typescript
   * await TurboPartner.removeUserFromPartnerPortal('user-uuid');
   * ```
   */
  static async removeUserFromPartnerPortal(userId: string): Promise<SuccessResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.delete<SuccessResponse>(`/partner/${partnerId}/users/${userId}`);
  }

  /**
   * Resend a partner portal invitation email to a user
   *
   * @param userId - User ID
   * @returns Success response
   *
   * @example
   * ```typescript
   * await TurboPartner.resendPartnerPortalInvitationToUser('user-uuid');
   * ```
   */
  static async resendPartnerPortalInvitationToUser(userId: string): Promise<SuccessResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.post<SuccessResponse>(`/partner/${partnerId}/users/${userId}/resend-invitation`);
  }

  // ============================================
  // AUDIT LOGS
  // ============================================

  /**
   * Get partner audit logs
   *
   * @param request - Optional filtering and pagination parameters
   * @returns Paginated list of audit log entries
   *
   * @example
   * ```typescript
   * const logs = await TurboPartner.getPartnerAuditLogs({
   *   action: 'org.created',
   *   startDate: '2025-01-01',
   *   endDate: '2025-12-31',
   * });
   * ```
   */
  static async getPartnerAuditLogs(request?: ListAuditLogsRequest): Promise<AuditLogListResponse> {
    const client = this.getClient();
    const partnerId = this.getPartnerId();
    return client.get<AuditLogListResponse>(`/partner/${partnerId}/audit-logs`, toQueryParams(request));
  }
}

File: packages/cli/cmd/partner/partner.go
package partner

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	turbodocx "github.com/TurboDocx/SDK/packages/go-sdk"
	"github.com/TurboDocx/SDK/packages/cli/cmd/cmdutil"
	"github.com/TurboDocx/SDK/packages/cli/internal/config"
	"github.com/spf13/cobra"
)

// PartnerClient defines the interface for TurboPartner operations (mockable in tests)
type PartnerClient interface {
	// Organization CRUD
	CreateOrganization(ctx context.Context, req *turbodocx.CreateOrganizationRequest) (*turbodocx.OrganizationResponse, error)
	ListOrganizations(ctx context.Context, req *turbodocx.ListOrganizationsRequest) (*turbodocx.OrganizationListResponse, error)
	GetOrganizationDetails(ctx context.Context, orgID string) (*turbodocx.OrganizationDetailResponse, error)
	UpdateOrganizationInfo(ctx context.Context, orgID string, req *turbodocx.UpdateOrganizationRequest) (*turbodocx.OrganizationResponse, error)
	DeleteOrganization(ctx context.Context, orgID string) (*turbodocx.SuccessResponse, error)
	UpdateOrganizationEntitlements(ctx context.Context, orgID string, req *turbodocx.UpdateEntitlementsRequest) (*turbodocx.EntitlementsResponse, error)

	// Organization Users
	ListOrganizationUsers(ctx context.Context, orgID string, req *turbodocx.ListOrgUsersRequest) (*turbodocx.OrgUserListResponse, error)
	AddUserToOrganization(ctx context.Context, orgID string, req *turbodocx.AddOrgUserRequest) (*turbodocx.OrgUserResponse, error)
	UpdateOrganizationUserRole(ctx context.Context, orgID, userID string, req *turbodocx.UpdateOrgUserRequest) (*turbodocx.OrgUserResponse, error)
	RemoveUserFromOrganization(ctx context.Context, orgID, userID string) (*turbodocx.SuccessResponse, error)
	ResendOrganizationInvitationToUser(ctx context.Context, orgID, userID string) (*turbodocx.SuccessResponse, error)

	// Organization API Keys
	ListOrganizationApiKeys(ctx context.Context, orgID string, req *turbodocx.ListOrgApiKeysRequest) (*turbodocx.OrgApiKeyListResponse, error)
	CreateOrganizationApiKey(ctx context.Context, orgID string, req *turbodocx.CreateOrgApiKeyRequest) (*turbodocx.OrgApiKeyResponse, error)
	UpdateOrganizationApiKey(ctx context.Context, orgID, keyID string, req *turbodocx.UpdateOrgApiKeyRequest) (*turbodocx.OrgApiKeyUpdateResponse, error)
	RevokeOrganizationApiKey(ctx context.Context, orgID, keyID string) (*turbodocx.SuccessResponse, error)

	// Partner API Keys
	ListPartnerApiKeys(ctx context.Context, req *turbodocx.ListPartnerApiKeysRequest) (*turbodocx.PartnerApiKeyListResponse, error)
	CreatePartnerApiKey(ctx context.Context, req *turbodocx.CreatePartnerApiKeyRequest) (*turbodocx.PartnerApiKeyResponse, error)
	UpdatePartnerApiKey(ctx context.Context, keyID string, req *turbodocx.UpdatePartnerApiKeyRequest) (*turbodocx.PartnerApiKeyUpdateResponse, error)
	RevokePartnerApiKey(ctx context.Context, keyID string) (*turbodocx.SuccessResponse, error)

	// Partner Users
	ListPartnerPortalUsers(ctx context.Context, req *turbodocx.ListPartnerUsersRequest) (*turbodocx.PartnerUserListResponse, error)
	AddUserToPartnerPortal(ctx context.Context, req *turbodocx.AddPartnerUserRequest) (*turbodocx.PartnerUserResponse, error)
	UpdatePartnerUserPermissions(ctx context.Context, userID string, req *turbodocx.UpdatePartnerUserRequest) (*turbodocx.PartnerUserUpdateResponse, error)
	RemoveUserFromPartnerPortal(ctx context.Context, userID string) (*turbodocx.SuccessResponse, error)
	ResendPartnerPortalInvitationToUser(ctx context.Context, userID string) (*turbodocx.SuccessResponse, error)

	// Audit Logs
	GetPartnerAuditLogs(ctx context.Context, req *turbodocx.ListAuditLogsRequest) (*turbodocx.AuditLogListResponse, error)
}

// PartnerClientFactory creates a PartnerClient from config. Package-level var for test injection.
type PartnerClientFactory func(cfg *config.Config) (PartnerClient, error)

// newPartnerClient creates a real SDK client. Replaceable in tests.
var newPartnerClient PartnerClientFactory = defaultNewPartnerClient

func defaultNewPartnerClient(cfg *config.Config) (PartnerClient, error) {
	if cfg.PartnerAPIKey == "" {
		return nil, fmt.Errorf("Partner API key is required. Run 'turbodocx config set partnerApiKey <key>' or pass --partner-api-key")
	}
	if cfg.PartnerID == "" {
		return nil, fmt.Errorf("Partner ID is required. Run 'turbodocx config set partnerId <id>' or pass --partner-id")
	}

	client, err := turbodocx.NewPartnerClient(turbodocx.PartnerConfig{
		PartnerAPIKey: cfg.PartnerAPIKey,
		PartnerID:     cfg.PartnerID,
		BaseURL:       cfg.BaseURL,
	})
	if err != nil {
		return nil, err
	}
	return client, nil
}

// parseJSONOrFile parses a string as JSON, or reads from a file if it starts with @
func parseJSONOrFile(input string, target interface{}) error {
	var data []byte
	if strings.HasPrefix(input, "@") {
		filePath := input[1:]
		var err error
		data, err = os.ReadFile(filePath)
		if err != nil {
			return fmt.Errorf("failed to read file %s: %w", filePath, err)
		}
	} else {
		data = []byte(input)
	}
	return json.Unmarshal(data, target)
}

// jsonMode is a package-level var for test control. In production, use cmdutil.IsJSON().
var jsonMode bool

func isJSONMode() bool {
	return jsonMode || cmdutil.IsJSON()
}

// PartnerCmd is the parent partner command
var PartnerCmd = &cobra.Command{
	Use:   "partner",
	Short: "Partner portal management",
	Long:  "Manage TurboPartner organizations, users, API keys, entitlements, and audit logs.",
}

// orgCmd is the parent org subcommand
var orgCmd = &cobra.Command{
	Use:   "org",
	Short: "Organization management",
}

// orgUserCmd is the parent org user subcommand
var orgUserCmd = &cobra.Command{
	Use:   "user",
	Short: "Organization user management",
}

// orgApikeyCmd is the parent org apikey subcommand
var orgApikeyCmd = &cobra.Command{
	Use:   "apikey",
	Short: "Organization API key management",
}

// apikeyCmd is the parent partner apikey subcommand
var apikeyCmd = &cobra.Command{
	Use:   "apikey",
	Short: "Partner API key management",
}

// userCmd is the parent partner user subcommand
var userCmd = &cobra.Command{
	Use:   "user",
	Short: "Partner portal user management",
}

// auditCmd is the parent audit subcommand
var auditCmd = &cobra.Command{
	Use:   "audit",
	Short: "Audit log management",
}

func init() {
	PartnerCmd.AddCommand(orgCmd)
	PartnerCmd.AddCommand(apikeyCmd)
	PartnerCmd.AddCommand(userCmd)
	PartnerCmd.AddCommand(auditCmd)

	orgCmd.AddCommand(orgUserCmd)
	orgCmd.AddCommand(orgApikeyCmd)
}
